File: backend/src/config/tracing.ts
import {
  Context,
  Span,
  SpanKind,
  SpanStatusCode,
  trace,
  context,
  propagation,
} from '@opentelemetry/api';
import { SpanExporter } from '@opentelemetry/sdk-trace-base';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { JaegerExporter } from '@opentelemetry/exporter-jaeger';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { ZipkinExporter } from '@opentelemetry/exporter-zipkin';
import { Resource } from '@opentelemetry/resources';
import { NodeSDK } from '@opentelemetry/sdk-node';
import {
  SEMRESATTRS_DEPLOYMENT_ENVIRONMENT,
  SEMRESATTRS_SERVICE_NAME,
  SEMRESATTRS_SERVICE_VERSION,
} from '@opentelemetry/semantic-conventions';

import { config } from '../config';
import { logger } from './logger';

/**
 * Configuration for distributed tracing.
 */
export interface TracingConfig {
  enabled: boolean;
  serviceName: string;
  serviceVersion: string;
  exporter: 'jaeger' | 'zipkin' | 'otlp' | 'none';
  sampleRate: number;
  jaegerEndpoint?: string;
  zipkinEndpoint?: string;
  otlpEndpoint?: string;
  autoInstrumentations: boolean;
  instrumentations: string[];
}

/**
 * Trace context carrier for propagating trace information across service boundaries.
 */
export interface TraceContext {
  traceId?: string;
  spanId?: string;
  traceFlags?: number;
  traceState?: string;
}

/**
 * Options for creating a traced function wrapper.
 */
export interface TracedFunctionOptions {
  name: string;
  kind?: SpanKind;
  attributes?: Record<string, string | number | boolean>;
}

let tracingInitialized = false;
let sdk: NodeSDK | null = null;
let tracer: ReturnType<typeof trace.getTracer> | null = null;

/**
 * Get the OpenTelemetry tracer instance.
 * Creates a new tracer if tracing is enabled, otherwise returns a no-op tracer.
 */
export function getTracer(): ReturnType<typeof trace.getTracer> {
  if (!tracer) {
    tracer = trace.getTracer(config.tracing.serviceName, config.tracing.serviceVersion);
  }
  return tracer;
}

/**
 * Extract trace context from HTTP headers (W3C Trace Context format).
 * This enables distributed trace context propagation across service calls.
 */
export function extractTraceContext(headers: Record<string, string | string[] | undefined>): {
  extractedContext: Context;
} {
  try {
    // Use OpenTelemetry's built-in W3C Trace Context propagation
    const carrier: Record<string, string> = {};

    // Convert headers to simple string record
    for (const [key, value] of Object.entries(headers)) {
      if (Array.isArray(value)) {
        carrier[key] = value[0];
      } else if (value) {
        carrier[key] = value;
      }
    }

    const extracted = propagation.extract(context.active(), carrier);
    return { extractedContext: extracted };
  } catch (error) {
    logger.warn('[Tracing] Failed to extract trace context', { error });
    return { extractedContext: context.active() };
  }
}

/**
 * Inject current trace context into headers for outgoing requests.
 * Uses W3C Trace Context format for standardized propagation.
 */
export function injectTraceContext(headers: Record<string, string>): void {
  try {
    propagation.inject(context.active(), headers);
  } catch (error) {
    logger.warn('[Tracing] Failed to inject trace context', { error });
  }
}

/**
 * Start a new span with optional parent context.
 */
export function startSpan(
  name: string,
  options?: {
    kind?: SpanKind;
    attributes?: Record<string, string | number | boolean>;
  }
): Span {
  const tracer = getTracer();

  return tracer.startSpan(name, {
    kind: options?.kind || SpanKind.INTERNAL,
    attributes: options?.attributes || {},
  });
}

/**
 * Execute a function within a traced span.
 * Automatically handles span creation, status, and end.
 */
export async function traceAsync<T>(
  name: string,
  fn: (span: Span) => Promise<T>,
  options?: {
    kind?: SpanKind;
    attributes?: Record<string, string | number | boolean>;
  }
): Promise<T> {
  const span = startSpan(name, options);

  try {
    const result = await fn(span);
    span.setStatus({ code: SpanStatusCode.OK });
    return result;
  } catch (error) {
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: error instanceof Error ? error.message : String(error),
    });
    span.recordException(error instanceof Error ? error : new Error(String(error)));
    throw error;
  } finally {
    span.end();
  }
}

/**
 * Execute a synchronous function within a traced span.
 */
export function traceSync<T>(
  name: string,
  fn: (span: Span) => T,
  options?: {
    kind?: SpanKind;
    attributes?: Record<string, string | number | boolean>;
  }
): T {
  const span = startSpan(name, options);

  try {
    const result = fn(span);
    span.setStatus({ code: SpanStatusCode.OK });
    return result;
  } catch (error) {
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: error instanceof Error ? error.message : String(error),
    });
    span.recordException(error instanceof Error ? error : new Error(String(error)));
    throw error;
  } finally {
    span.end();
  }
}

/**
 * Add event to current span.
 */
export function addSpanEvent(
  name: string,
  attributes?: Record<string, string | number | boolean>
): void {
  const activeSpan = trace.getSpan(context.active());
  if (activeSpan) {
    activeSpan.addEvent(name, attributes);
  }
}

/**
 * Set attribute on current span.
 */
export function setSpanAttribute(key: string, value: string | number | boolean): void {
  const activeSpan = trace.getSpan(context.active());
  if (activeSpan) {
    activeSpan.setAttribute(key, value);
  }
}

/**
 * Wrap an RPC handler with tracing.
 * Automatically creates spans with relevant Nakama context information.
 */
export function wrapRpcHandler<T>(
  handlerName: string,
  handler: (ctx: unknown, logger: unknown, nk: unknown, payload: string) => Promise<T> | T
): (ctx: unknown, logger: unknown, nk: unknown, payload: string) => Promise<T> | T {
  return async (ctx: unknown, logger: unknown, nk: unknown, payload: string) => {
    const span = startSpan(`rpc.${handlerName}`, {
      kind: SpanKind.SERVER,
      attributes: {
        'rpc.system': 'nakama',
        'rpc.method': handlerName,
        'deployment.environment': config.environment,
      },
    });

    // Add context attributes if available
    const nakamaCtx = ctx as { userId?: string; sessionId?: string; matchId?: string } | null;
    if (nakamaCtx?.userId) {
      span.setAttribute('user.id', nakamaCtx.userId);
    }
    if (nakamaCtx?.sessionId) {
      span.setAttribute('session.id', nakamaCtx.sessionId);
    }
    if (nakamaCtx?.matchId) {
      span.setAttribute('match.id', nakamaCtx.matchId);
    }

    try {
      const result = await handler(ctx, logger, nk, payload);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (error) {
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: error instanceof Error ? error.message : String(error),
      });
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      throw error;
    } finally {
      span.end();
    }
  };
}

/**
 * Initialize the OpenTelemetry tracing SDK.
 */
export function initializeTracing(): void {
  const tracingConfig = config.tracing;

  if (!tracingConfig.enabled) {
    logger.info('[Tracing] Distributed tracing is disabled');
    return;
  }

  if (tracingInitialized) {
    logger.warn('[Tracing] Tracing already initialized');
    return;
  }

  try {
    // Create resource with service information
    const resource = new Resource({
      [SEMRESATTRS_SERVICE_NAME]: tracingConfig.serviceName,
      [SEMRESATTRS_SERVICE_VERSION]: tracingConfig.serviceVersion,
      [SEMRESATTRS_DEPLOYMENT_ENVIRONMENT]: config.environment,
    });

    // Set up exporters based on configuration
    let exporter;
    switch (tracingConfig.exporter) {
      case 'jaeger':
        exporter = new JaegerExporter({
          endpoint: tracingConfig.jaegerEndpoint || 'http://localhost:14268',
        });
        logger.info(`[Tracing] Jaeger exporter configured: ${tracingConfig.jaegerEndpoint}`);
        break;
      case 'zipkin':
        exporter = new ZipkinExporter({
          url: tracingConfig.zipkinEndpoint || 'http://localhost:9411',
        });
        logger.info(`[Tracing] Zipkin exporter configured: ${tracingConfig.zipkinEndpoint}`);
        break;
      case 'otlp':
        exporter = new OTLPTraceExporter({
          url: tracingConfig.otlpEndpoint || 'http://localhost:4318',
        });
        logger.info(`[Tracing] OTLP exporter configured: ${tracingConfig.otlpEndpoint}`);
        break;
      case 'none':
        logger.info('[Tracing] Tracing initialized without exporter (for development)');
        break;
      default:
        logger.warn(`[Tracing] Unknown exporter type: ${tracingConfig.exporter}`);
        break;
    }

    // Build instrumentations list
    let instrumentations: unknown[] = [];

    if (tracingConfig.autoInstrumentations) {
      const autoInst = getNodeAutoInstrumentations({
        // Only enable specified instrumentations
        '@opentelemetry/instrumentation-http': {
          enabled: tracingConfig.instrumentations.includes('http'),
        },
        '@opentelemetry/instrumentation-express': {
          enabled: tracingConfig.instrumentations.includes('express'),
        },
        '@opentelemetry/instrumentation-pg': {
          enabled: tracingConfig.instrumentations.includes('pg'),
        },
      });
      // getNodeAutoInstrumentations returns an array
      instrumentations = autoInst as unknown[];
    }

    // Create and start the SDK
    // Note: Type assertion needed due to version mismatch between exporters in dependency tree
    sdk = new NodeSDK({
      resource,
      traceExporter: exporter as any,
      instrumentations: instrumentations as never[],
      serviceName: tracingConfig.serviceName,
    });

    // Start the SDK
    sdk.start();
    logger.info(
      `[Tracing] OpenTelemetry SDK started - service: ${tracingConfig.serviceName}, environment: ${config.environment}`
    );

    tracingInitialized = true;

    // Handle graceful shutdown
    process.on('SIGTERM', () => {
      logger.info('[Tracing] Shutting down OpenTelemetry SDK');
      sdk?.shutdown().catch((err) => {
        logger.error('[Tracing] Error shutting down SDK', err);
      });
    });
  } catch (error) {
    logger.error('[Tracing] Failed to initialize tracing', error);
  }
}

/**
 * Get the current tracing configuration.
 */
export function getTracingConfig(): TracingConfig {
  return config.tracing;
}

/**
 * Check if tracing is enabled.
 */
export function isTracingEnabled(): boolean {
  return config.tracing.enabled && tracingInitialized;
}

/**
 * Shutdown the tracing SDK.
 */
export async function shutdownTracing(): Promise<void> {
  if (sdk) {
    await sdk.shutdown();
    tracingInitialized = false;
    logger.info('[Tracing] Tracing SDK shut down');
  }
}

File: backend/metrics/tracing.go
// Package metrics provides OpenTelemetry tracing for the Armored Archer backend.
// This package implements distributed tracing using OpenTelemetry and Grafana Tempo.
package metrics

import (
	"context"
	"fmt"
	"os"
	"sync"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"go.opentelemetry.io/otel/trace"
	"go.opentelemetry.io/otel/trace/noop"
)

// TracingConfig holds tracing configuration.
type TracingConfig struct {
	// Enabled enables/disables tracing
	Enabled bool `json:"enabled" yaml:"enabled"`
	// ServiceName is the name of the service
	ServiceName string `json:"service_name" yaml:"service_name"`
	// ServiceVersion is the version of the service
	ServiceVersion string `json:"service_version" yaml:"service_version"`
	// OTLPEndpoint is the OTLP collector endpoint (host:port)
	OTLPEndpoint string `json:"otlp_endpoint" yaml:"otlp_endpoint"`
	// SamplingRate is the sampling rate (0.0-1.0)
	// For production, use 0.01 (1%) to reduce overhead
	// For development, use 1.0 (100%) for full tracing
	SamplingRate float64 `json:"sampling_rate" yaml:"sampling_rate"`
	// BatchTimeout is the timeout for batch exporting
	BatchTimeout time.Duration `json:"batch_timeout" yaml:"batch_timeout"`
	// MaxExportBatchSize is the maximum batch size for exporting
	MaxExportBatchSize int `json:"max_export_batch_size" yaml:"max_export_batch_size"`
	// Environment is the deployment environment
	Environment string `json:"environment" yaml:"environment"`
}

// DefaultTracingConfig returns default tracing configuration.
func DefaultTracingConfig() *TracingConfig {
	return &TracingConfig{
		Enabled:            true,
		ServiceName:        "armored-archer-backend",
		ServiceVersion:     "v2.1.0",
		OTLPEndpoint:       "otel-collector:4317",
		SamplingRate:       0.01, // 1% sampling for production
		BatchTimeout:       5 * time.Second,
		MaxExportBatchSize: 512,
		Environment:        getEnvOrDefault("ENVIRONMENT", "development"),
	}
}

// TracingProvider holds the tracing provider and related components.
type TracingProvider struct {
	config     *TracingConfig
	tp         trace.TracerProvider
	shutdownFn func(context.Context) error
	mu         sync.RWMutex
	tracer     trace.Tracer
}

// globalTracingProvider is the global tracing provider instance.
var globalTracingProvider *TracingProvider
var tracingOnce sync.Once

// GetTracingProvider returns the global tracing provider.
func GetTracingProvider() *TracingProvider {
	tracingOnce.Do(func() {
		globalTracingProvider = NewTracingProvider(DefaultTracingConfig())
	})
	return globalTracingProvider
}

// NewTracingProvider creates a new tracing provider.
func NewTracingProvider(config *TracingConfig) *TracingProvider {
	tp := &TracingProvider{
		config: config,
	}

	if config.Enabled {
		if err := tp.initialize(); err != nil {
			// Log error but continue with noop tracer
			fmt.Printf("Warning: Failed to initialize tracing: %v\n", err)
			tp.setupNoop()
		}
	} else {
		tp.setupNoop()
	}

	return tp
}

// initialize sets up the OpenTelemetry tracing provider.
func (tp *TracingProvider) initialize() error {
	tp.mu.Lock()
	defer tp.mu.Unlock()

	// Create OTLP exporter
	exporter, err := otlptrace.New(
		context.Background(),
		otlptracegrpc.NewClient(
			otlptracegrpc.WithEndpoint(tp.config.OTLPEndpoint),
			otlptracegrpc.WithInsecure(), // Use TLS in production
		),
	)
	if err != nil {
		return fmt.Errorf("failed to create OTLP exporter: %w", err)
	}

	// Create resource with service attributes
	res, err := resource.New(
		context.Background(),
		semconv.ServiceName(tp.config.ServiceName),
		semconv.ServiceVersion(tp.config.ServiceVersion),
		semconv.DeploymentEnvironment(tp.config.Environment),
		attribute.String("service.instance.id", getInstanceID()),
		attribute.String("host.name", getHostname()),
	)
	if err != nil {
		return fmt.Errorf("failed to create resource: %w", err)
	}

	// Create sampler based on sampling rate
	sampler := sdktrace.ParentBased(
		sdktrace.TraceIDRatioBased(tp.config.SamplingRate),
	)

	// Create trace provider
	traceProvider := sdktrace.NewTracerProvider(
		sdktrace.WithSampler(sampler),
		sdktrace.WithResource(res),
		sdktrace.WithBatcher(
			exporter,
			sdktrace.WithBatchTimeout(tp.config.BatchTimeout),
			sdktrace.WithMaxExportBatchSize(tp.config.MaxExportBatchSize),
		),
	)

	// Set global tracer provider
	otel.SetTracerProvider(traceProvider)

	// Set global propagator (W3C Trace Context + Baggage)
	otel.SetTextMapPropagator(
		propagation.NewCompositeTextMapPropagator(
			propagation.TraceContext{},
			propagation.Baggage{},
		),
	)

	tp.tp = traceProvider
	tp.tracer = traceProvider.Tracer(tp.config.ServiceName)
	tp.shutdownFn = func(ctx context.Context) error {
		return traceProvider.Shutdown(ctx)
	}

	fmt.Printf("Tracing initialized: service=%s, sampling_rate=%.2f%%, endpoint=%s\n",
		tp.config.ServiceName,
		tp.config.SamplingRate*100,
		tp.config.OTLPEndpoint,
	)

	return nil
}

// setupNoop sets up a no-op tracer for disabled tracing.
func (tp *TracingProvider) setupNoop() {
	tp.mu.Lock()
	defer tp.mu.Unlock()

	tp.tp = noop.NewTracerProvider()
	tp.tracer = tp.tp.Tracer("noop")
	tp.shutdownFn = func(ctx context.Context) error { return nil }
}

// Tracer returns the tracer for the given name.
func (tp *TracingProvider) Tracer(name string) trace.Tracer {
	tp.mu.RLock()
	defer tp.mu.RUnlock()

	if name == "" {
		return tp.tracer
	}
	return tp.tp.Tracer(name)
}

// Shutdown gracefully shuts down the tracing provider.
func (tp *TracingProvider) Shutdown(ctx context.Context) error {
	tp.mu.Lock()
	defer tp.mu.Unlock()

	if tp.shutdownFn != nil {
		return tp.shutdownFn(ctx)
	}
	return nil
}

// StartSpan starts a new span with the given name and options.
func (tp *TracingProvider) StartSpan(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
	tp.mu.RLock()
	tracer := tp.tracer
	tp.mu.RUnlock()

	return tracer.Start(ctx, spanName, opts...)
}

// IsTracingEnabled returns whether tracing is enabled.
func (tp *TracingProvider) IsTracingEnabled() bool {
	return tp.config.Enabled
}

// GetConfig returns the tracing configuration.
func (tp *TracingProvider) GetConfig() *TracingConfig {
	return tp.config
}

// SpanTimer is a helper for timing spans.
type SpanTimer struct {
	ctx  context.Context
	span trace.Span
}

// Stop ends the span and records the duration.
func (st *SpanTimer) Stop() {
	if st.span != nil {
		st.span.End()
	}
}

// SetStatus sets the span status.
func (st *SpanTimer) SetStatus(code trace.StatusCode, description string) {
	if st.span != nil {
		st.span.SetStatus(code, description)
	}
}

// RecordError records an error in the span.
func (st *SpanTimer) RecordError(err error) {
	if st.span != nil {
		st.span.RecordError(err)
	}
}

// SetAttribute sets an attribute on the span.
func (st *SpanTimer) SetAttribute(key string, value interface{}) {
	if st.span != nil {
		switch v := value.(type) {
		case string:
			st.span.SetAttributes(attribute.String(key, v))
		case int:
			st.span.SetAttributes(attribute.Int(key, v))
		case int64:
			st.span.SetAttributes(attribute.Int64(key, v))
		case float64:
			st.span.SetAttributes(attribute.Float64(key, v))
		case bool:
			st.span.SetAttributes(attribute.Bool(key, v))
		case []string:
			st.span.SetAttributes(attribute.StringSlice(key, v))
		default:
			st.span.SetAttributes(attribute.String(key, fmt.Sprintf("%v", v)))
		}
	}
}

// StartTimedSpan starts a span and returns a timer for automatic duration tracking.
func (tp *TracingProvider) StartTimedSpan(ctx context.Context, spanName string) *SpanTimer {
	ctx, span := tp.StartSpan(ctx, spanName)
	return &SpanTimer{
		ctx:  ctx,
		span: span,
	}
}

// GetTraceID extracts the trace ID from the context.
func GetTraceID(ctx context.Context) string {
	spanContext := trace.SpanFromContext(ctx).SpanContext()
	if spanContext.HasTraceID() {
		return spanContext.TraceID().String()
	}
	return ""
}

// GetSpanID extracts the span ID from the context.
func GetSpanID(ctx context.Context) string {
	spanContext := trace.SpanFromContext(ctx).SpanContext()
	if spanContext.HasSpanID() {
		return spanContext.SpanID().String()
	}
	return ""
}

// IsSampled returns true if the current span is sampled.
func IsSampled(ctx context.Context) bool {
	return trace.SpanFromContext(ctx).SpanContext().IsSampled()
}

// Inject injects tracing context into a carrier (for outgoing requests).
func Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
	otel.GetTextMapPropagator().Inject(ctx, carrier)
}

// Extract extracts tracing context from a carrier (for incoming requests).
func Extract(ctx context.Context, carrier propagation.TextMapCarrier) context.Context {
	return otel.GetTextMapPropagator().Extract(ctx, carrier)
}

// TextMapCarrier is a simple text map carrier for propagation.
type TextMapCarrier map[string]string

// Get returns a value from the carrier.
func (c TextMapCarrier) Get(key string) string {
	val, ok := c[key]
	if !ok {
		return ""
	}
	return val
}

// Set sets a value in the carrier.
func (c TextMapCarrier) Set(key string, value string) {
	c[key] = value
}

// Keys returns all keys in the carrier.
func (c TextMapCarrier) Keys() []string {
	keys := make([]string, 0, len(c))
	for k := range c {
		keys = append(keys, k)
	}
	return keys
}

// Common attribute keys for tracing.
const (
	AttributeRPCName      = "rpc.name"
	AttributeRPCMethod    = "rpc.method"
	AttributeRPCSystem    = "rpc.system"
	AttributeUserID       = "user.id"
	AttributePlayerID     = "player.id"
	AttributeMatchID      = "match.id"
	AttributeMatchType    = "match.type"
	AttributeCombatAction = "combat.action"
	AttributeGearType     = "gear.type"
	AttributeGearRarity   = "gear.rarity"
	AttributeCurrencyType = "currency.type"
	AttributeSeason       = "season"
	AttributeErrorCode    = "error.code"
	AttributeErrorMessage = "error.message"
	AttributeDBQuery      = "db.query"
	AttributeDBSystem     = "db.system"
	AttributeCacheKey     = "cache.key"
	AttributeCacheHit     = "cache.hit"
	AttributeHTTPMethod   = "http.method"
	AttributeHTTPURL      = "http.url"
	AttributeHTTPStatus   = "http.status_code"
)

// TraceRPC wraps an RPC call with tracing.
func TraceRPC(ctx context.Context, rpcName string, userID string, fn func(context.Context) (string, error)) (string, error) {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, fmt.Sprintf("rpc/%s", rpcName))
	defer timer.Stop()

	// Set common attributes
	timer.SetAttribute(AttributeRPCName, rpcName)
	timer.SetAttribute(AttributeRPCSystem, "nakama")
	if userID != "" {
		timer.SetAttribute(AttributeUserID, userID)
	}

	// Execute the RPC
	result, err := fn(timer.ctx)

	// Record error if present
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
		timer.SetAttribute(AttributeErrorCode, "rpc_error")
	}

	return result, err
}

// TraceDatabase wraps a database operation with tracing.
func TraceDatabase(ctx context.Context, query string, fn func(context.Context) error) error {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, "db/query")
	defer timer.Stop()

	timer.SetAttribute(AttributeDBQuery, truncateString(query, 500))
	timer.SetAttribute(AttributeDBSystem, "postgresql")

	err := fn(timer.ctx)
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// TraceCache wraps a cache operation with tracing.
func TraceCache(ctx context.Context, operation string, key string, hit bool, fn func(context.Context) error) error {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, fmt.Sprintf("cache/%s", operation))
	defer timer.Stop()

	timer.SetAttribute(AttributeCacheKey, key)
	timer.SetAttribute(AttributeCacheHit, hit)

	err := fn(timer.ctx)
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// TraceMatch wraps a match operation with tracing.
func TraceMatch(ctx context.Context, matchType string, matchID string, fn func(context.Context) error) error {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, fmt.Sprintf("match/%s", matchType))
	defer timer.Stop()

	timer.SetAttribute(AttributeMatchType, matchType)
	if matchID != "" {
		timer.SetAttribute(AttributeMatchID, matchID)
	}

	err := fn(timer.ctx)
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// TraceCombat wraps a combat action with tracing.
func TraceCombat(ctx context.Context, actionType string, fn func(context.Context) error) error {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, fmt.Sprintf("combat/%s", actionType))
	defer timer.Stop()

	timer.SetAttribute(AttributeCombatAction, actionType)

	err := fn(timer.ctx)
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// TraceGear wraps a gear operation with tracing.
func TraceGear(ctx context.Context, operation string, gearType string, rarity string, fn func(context.Context) error) error {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, fmt.Sprintf("gear/%s", operation))
	defer timer.Stop()

	timer.SetAttribute(AttributeGearType, gearType)
	timer.SetAttribute(AttributeGearRarity, rarity)

	err := fn(timer.ctx)
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// TraceSeason wraps a season operation with tracing.
func TraceSeason(ctx context.Context, operation string, season string, fn func(context.Context) error) error {
	tp := GetTracingProvider()
	timer := tp.StartTimedSpan(ctx, fmt.Sprintf("season/%s", operation))
	defer timer.Stop()

	timer.SetAttribute(AttributeSeason, season)

	err := fn(timer.ctx)
	if err != nil {
		timer.RecordError(err)
		timer.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// Helper functions

func getEnvOrDefault(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getInstanceID() string {
	if id := os.Getenv("INSTANCE_ID"); id != "" {
		return id
	}
	if id := os.Getenv("HOSTNAME"); id != "" {
		return id
	}
	return "unknown"
}

func getHostname() string {
	if hostname := os.Getenv("HOSTNAME"); hostname != "" {
		return hostname
	}
	return "unknown"
}

func truncateString(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}

// InitializeTracing initializes the global tracing provider.
// Call this once at application startup.
func InitializeTracing(config *TracingConfig) error {
	if config == nil {
		config = DefaultTracingConfig()
	}

	// Override global provider
	globalTracingProvider = NewTracingProvider(config)
	return nil
}

// ShutdownTracing gracefully shuts down tracing.
// Call this at application shutdown.
func ShutdownTracing(ctx context.Context) error {
	if globalTracingProvider != nil {
		return globalTracingProvider.Shutdown(ctx)
	}
	return nil
}

// GetTracer returns the global tracer.
func GetTracer() trace.Tracer {
	return GetTracingProvider().Tracer("")
}
